File: sdks/go/timing/analyzer.go
package timing

import (
	"fmt"
	"math"

	xagentauth "github.com/dyshay/agentauth/sdks/go"
)

// AnalyzeParams contains the parameters for timing analysis.
type AnalyzeParams struct {
	ElapsedMs     float64
	ChallengeType string
	Difficulty    xagentauth.Difficulty
	RTTMs         float64
}

// TimingAnalyzer classifies response timing into zones and computes
// penalties, z-scores, and confidence levels.
type TimingAnalyzer struct {
	baselines map[string]xagentauth.TimingBaseline
	defaults  struct {
		tooFast float64
		aiLower float64
		aiUpper float64
		human   float64
		timeout float64
	}
}

// NewTimingAnalyzer creates a new TimingAnalyzer from the given config.
func NewTimingAnalyzer(config *xagentauth.TimingConfig) *TimingAnalyzer {
	ta := &TimingAnalyzer{
		baselines: make(map[string]xagentauth.TimingBaseline),
	}

	allBaselines := DefaultBaselines
	if config != nil && len(config.Baselines) > 0 {
		allBaselines = config.Baselines
	}
	for _, b := range allBaselines {
		key := fmt.Sprintf("%s:%s", b.ChallengeType, b.Difficulty)
		ta.baselines[key] = b
	}

	ta.defaults.tooFast = 50
	ta.defaults.aiLower = 50
	ta.defaults.aiUpper = 2000
	ta.defaults.human = 10000
	ta.defaults.timeout = 30000

	if config != nil {
		if config.DefaultTooFastMs > 0 {
			ta.defaults.tooFast = config.DefaultTooFastMs
		}
		if config.DefaultAILowerMs > 0 {
			ta.defaults.aiLower = config.DefaultAILowerMs
		}
		if config.DefaultAIUpperMs > 0 {
			ta.defaults.aiUpper = config.DefaultAIUpperMs
		}
		if config.DefaultHumanMs > 0 {
			ta.defaults.human = config.DefaultHumanMs
		}
		if config.DefaultTimeoutMs > 0 {
			ta.defaults.timeout = config.DefaultTimeoutMs
		}
	}

	return ta
}

// Analyze classifies the elapsed time into a timing zone and computes metrics.
func (ta *TimingAnalyzer) Analyze(params AnalyzeParams) xagentauth.TimingAnalysis {
	key := fmt.Sprintf("%s:%s", params.ChallengeType, params.Difficulty)
	baseline, ok := ta.baselines[key]
	if !ok {
		baseline = ta.makeDefaultBaseline()
	}

	// Apply RTT tolerance to zone boundaries
	tolerance := 0.0
	if params.RTTMs > 0 {
		tolerance = math.Max(params.RTTMs*0.5, 200)
	}
	adjustedBaseline := baseline
	if tolerance > 0 {
		adjustedBaseline.AIUpperMs = baseline.AIUpperMs + tolerance
		adjustedBaseline.HumanMs = baseline.HumanMs + tolerance
	}

	zone := ta.classifyZone(params.ElapsedMs, adjustedBaseline)
	penalty := ta.computePenalty(zone, params.ElapsedMs, adjustedBaseline)
	zScore := ta.computeZScore(params.ElapsedMs, baseline)
	confidence := ta.computeConfidence(params.ElapsedMs, adjustedBaseline, zone)
	details := ta.describeZone(zone, params.ElapsedMs, adjustedBaseline)

	// Round-number detection: flag suspiciously round elapsed times in ai_zone
	isRound := int(params.ElapsedMs)%500 == 0 || int(params.ElapsedMs)%100 == 0
	if isRound && zone == xagentauth.ZoneAI && params.ElapsedMs > 0 {
		confidence = math.Round(confidence*0.85*1000) / 1000
		details += " [round-number timing detected]"
	}

	return xagentauth.TimingAnalysis{
		ElapsedMs:  params.ElapsedMs,
		Zone:       string(zone),
		Confidence: confidence,
		ZScore:     math.Round(zScore*100) / 100,
		Penalty:    math.Round(penalty*1000) / 1000,
		Details:    details,
	}
}

// AnalyzePattern performs pattern analysis on a series of step timings.
func (ta *TimingAnalyzer) AnalyzePattern(stepTimings []float64) xagentauth.TimingPatternAnalysis {
	if len(stepTimings) < 2 {
		return xagentauth.TimingPatternAnalysis{
			VarianceCoefficient: 0,
			Trend:               "constant",
			RoundNumberRatio:    0,
			Verdict:             "inconclusive",
		}
	}

	mean := 0.0
	for _, t := range stepTimings {
		mean += t
	}
	mean /= float64(len(stepTimings))

	variance := 0.0
	for _, t := range stepTimings {
		diff := t - mean
		variance += diff * diff
	}
	variance /= float64(len(stepTimings))
	std := math.Sqrt(variance)

	vc := 0.0
	if mean > 0 {
		vc = std / mean
	}

	trend := ta.detectTrend(stepTimings)

	// Round number detection
	roundCount := 0
	for _, t := range stepTimings {
		ti := int(t)
		if ti%500 == 0 || (ti%100 == 0 && ti%500 != 0) {
			roundCount++
		}
	}
	roundRatio := float64(roundCount) / float64(len(stepTimings))

	// Verdict
	verdict := "inconclusive"
	if vc < 0.05 && len(stepTimings) >= 3 {
		verdict = "artificial"
	} else if roundRatio > 0.5 {
		verdict = "artificial"
	} else if vc > 0.1 {
		verdict = "natural"
	}

	return xagentauth.TimingPatternAnalysis{
		VarianceCoefficient: math.Round(vc*1000) / 1000,
		Trend:               trend,
		RoundNumberRatio:    math.Round(roundRatio*100) / 100,
		Verdict:             verdict,
	}
}

func (ta *TimingAnalyzer) makeDefaultBaseline() xagentauth.TimingBaseline {
	return xagentauth.TimingBaseline{
		ChallengeType: "default",
		Difficulty:    xagentauth.DifficultyMedium,
		MeanMs:        (ta.defaults.aiLower + ta.defaults.aiUpper) / 2,
		StdMs:         (ta.defaults.aiUpper - ta.defaults.aiLower) / 4,
		TooFastMs:     ta.defaults.tooFast,
		AILowerMs:     ta.defaults.aiLower,
		AIUpperMs:     ta.defaults.aiUpper,
		HumanMs:       ta.defaults.human,
		TimeoutMs:     ta.defaults.timeout,
	}
}

func (ta *TimingAnalyzer) classifyZone(elapsed float64, baseline xagentauth.TimingBaseline) xagentauth.TimingZone {
	if elapsed < baseline.TooFastMs {
		return xagentauth.ZoneTooFast
	}
	if elapsed >= baseline.TooFastMs && elapsed <= baseline.AIUpperMs {
		return xagentauth.ZoneAI
	}
	if elapsed > baseline.AIUpperMs && elapsed <= baseline.HumanMs {
		return xagentauth.ZoneSuspicious
	}
	if elapsed > baseline.HumanMs && elapsed <= baseline.TimeoutMs {
		return xagentauth.ZoneHuman
	}
	return xagentauth.ZoneTimeout
}

func (ta *TimingAnalyzer) computePenalty(zone xagentauth.TimingZone, elapsed float64, baseline xagentauth.TimingBaseline) float64 {
	switch zone {
	case xagentauth.ZoneTooFast:
		return 1.0
	case xagentauth.ZoneAI:
		return 0.0
	case xagentauth.ZoneSuspicious:
		rangeVal := baseline.HumanMs - baseline.AIUpperMs
		if rangeVal <= 0 {
			return 0.5
		}
		position := (elapsed - baseline.AIUpperMs) / rangeVal
		return 0.3 + position*0.4
	case xagentauth.ZoneHuman:
		return 0.9
	case xagentauth.ZoneTimeout:
		return 1.0
	}
	return 0.0
}

func (ta *TimingAnalyzer) computeZScore(elapsed float64, baseline xagentauth.TimingBaseline) float64 {
	if baseline.StdMs == 0 {
		return 0
	}
	return (elapsed - baseline.MeanMs) / baseline.StdMs
}

func (ta *TimingAnalyzer) computeConfidence(elapsed float64, baseline xagentauth.TimingBaseline, zone xagentauth.TimingZone) float64 {
	switch zone {
	case xagentauth.ZoneTooFast:
		ratio := elapsed / baseline.TooFastMs
		return math.Max(0.5, 1-ratio)
	case xagentauth.ZoneAI:
		distFromMean := math.Abs(elapsed - baseline.MeanMs)
		normalizedDist := distFromMean / baseline.StdMs
		return math.Max(0.5, math.Min(1, 1-normalizedDist*0.15))
	case xagentauth.ZoneSuspicious:
		rangeVal := baseline.HumanMs - baseline.AIUpperMs
		if rangeVal <= 0 {
			return 0.4
		}
		return 0.4 + 0.2*((elapsed-baseline.AIUpperMs)/rangeVal)
	case xagentauth.ZoneHuman:
		return 0.8
	case xagentauth.ZoneTimeout:
		return 0.95
	}
	return 0.5
}

func (ta *TimingAnalyzer) describeZone(zone xagentauth.TimingZone, elapsed float64, baseline xagentauth.TimingBaseline) string {
	ms := int(math.Round(elapsed))
	switch zone {
	case xagentauth.ZoneTooFast:
		return fmt.Sprintf("Response time %dms is below %.0fms threshold -- likely pre-computed or scripted", ms, baseline.TooFastMs)
	case xagentauth.ZoneAI:
		return fmt.Sprintf("Response time %dms is within expected AI range [%.0fms, %.0fms]", ms, baseline.AILowerMs, baseline.AIUpperMs)
	case xagentauth.ZoneSuspicious:
		return fmt.Sprintf("Response time %dms exceeds AI range -- possible human assistance", ms)
	case xagentauth.ZoneHuman:
		return fmt.Sprintf("Response time %dms exceeds %.0fms -- likely human solver", ms, baseline.HumanMs)
	case xagentauth.ZoneTimeout:
		return fmt.Sprintf("Response time %dms exceeds timeout threshold of %.0fms", ms, baseline.TimeoutMs)
	}
	return ""
}

func (ta *TimingAnalyzer) detectTrend(timings []float64) string {
	if len(timings) < 3 {
		return "variable"
	}

	n := float64(len(timings))
	xMean := (n - 1) / 2
	yMean := 0.0
	for _, t := range timings {
		yMean += t
	}
	yMean /= n

	numerator := 0.0
	denominator := 0.0
	for i, t := range timings {
		xi := float64(i) - xMean
		yi := t - yMean
		numerator += xi * yi
		denominator += xi * xi
	}

	if denominator == 0 {
		return "constant"
	}
	slope := numerator / denominator

	normalizedSlope := 0.0
	if yMean > 0 {
		normalizedSlope = slope / yMean
	}

	if math.Abs(normalizedSlope) < 0.05 {
		return "constant"
	}
	if normalizedSlope > 0.1 {
		return "increasing"
	}
	if normalizedSlope < -0.1 {
		return "decreasing"
	}
	return "variable"
}

File: sdks/rust/src/timing/analyzer.rs
use std::collections::HashMap;

use crate::types::{
    Difficulty, TimingAnalysis, TimingBaseline, TimingConfig, TimingPatternAnalysis, TimingTrend,
    TimingVerdict, TimingZone,
};

use super::baselines::default_baselines;

/// Analyzes response timing to classify solver type (AI, human, scripted).
pub struct TimingAnalyzer {
    baselines: HashMap<String, TimingBaseline>,
    defaults: DefaultThresholds,
}

struct DefaultThresholds {
    too_fast: f64,
    ai_lower: f64,
    ai_upper: f64,
    human: f64,
    timeout: f64,
}

impl TimingAnalyzer {
    pub fn new(config: &TimingConfig) -> Self {
        let mut baselines = HashMap::new();
        let all = config.baselines.as_ref().map(|b| b.as_slice()).unwrap_or(&[]);

        let source = if all.is_empty() {
            default_baselines()
        } else {
            all.to_vec()
        };

        for b in source {
            let key = format!("{}:{:?}", b.challenge_type, b.difficulty);
            baselines.insert(key, b);
        }

        Self {
            baselines,
            defaults: DefaultThresholds {
                too_fast: config.default_too_fast_ms,
                ai_lower: config.default_ai_lower_ms,
                ai_upper: config.default_ai_upper_ms,
                human: config.default_human_ms,
                timeout: config.default_timeout_ms,
            },
        }
    }

    /// Analyze a single response timing.
    pub fn analyze(
        &self,
        elapsed_ms: f64,
        challenge_type: &str,
        difficulty: &Difficulty,
        rtt_ms: Option<f64>,
    ) -> TimingAnalysis {
        let key = format!("{}:{:?}", challenge_type, difficulty);
        let baseline = self
            .baselines
            .get(&key)
            .cloned()
            .unwrap_or_else(|| self.make_default_baseline());

        // Apply RTT tolerance
        let tolerance = match rtt_ms {
            Some(rtt) if rtt > 0.0 => f64::max(rtt * 0.5, 200.0),
            _ => 0.0,
        };

        let adjusted = if tolerance > 0.0 {
            TimingBaseline {
                ai_upper_ms: baseline.ai_upper_ms + tolerance,
                human_ms: baseline.human_ms + tolerance,
                ..baseline.clone()
            }
        } else {
            baseline.clone()
        };

        let zone = self.classify_zone(elapsed_ms, &adjusted);
        let penalty = self.compute_penalty(&zone, elapsed_ms, &adjusted);
        let z_score = self.compute_z_score(elapsed_ms, &baseline);
        let mut confidence = self.compute_confidence(elapsed_ms, &adjusted, &zone);
        let mut details = self.describe_zone(&zone, elapsed_ms, &adjusted);

        // Round-number detection in ai_zone
        let is_round = elapsed_ms > 0.0
            && (elapsed_ms % 500.0 == 0.0 || elapsed_ms % 100.0 == 0.0);
        if is_round && zone == TimingZone::AiZone {
            confidence = (confidence * 0.85 * 1000.0).round() / 1000.0;
            details += " [round-number timing detected]";
        }

        TimingAnalysis {
            elapsed_ms,
            zone,
            confidence,
            z_score: (z_score * 100.0).round() / 100.0,
            penalty: (penalty * 1000.0).round() / 1000.0,
            details,
        }
    }

    /// Analyze per-step timing patterns.
    pub fn analyze_pattern(&self, step_timings: &[f64]) -> TimingPatternAnalysis {
        if step_timings.len() < 2 {
            return TimingPatternAnalysis {
                variance_coefficient: 0.0,
                trend: TimingTrend::Constant,
                round_number_ratio: 0.0,
                verdict: TimingVerdict::Inconclusive,
            };
        }

        let mean: f64 = step_timings.iter().sum::<f64>() / step_timings.len() as f64;
        let variance: f64 = step_timings.iter().map(|t| (t - mean).powi(2)).sum::<f64>()
            / step_timings.len() as f64;
        let std = variance.sqrt();
        let variance_coefficient = if mean > 0.0 { std / mean } else { 0.0 };

        let trend = self.detect_trend(step_timings);

        // Round number ratio
        let round_count = step_timings
            .iter()
            .filter(|t| **t % 500.0 == 0.0 || (**t % 100.0 == 0.0 && **t % 500.0 != 0.0))
            .count();
        let round_number_ratio = round_count as f64 / step_timings.len() as f64;

        // Verdict
        let verdict = if variance_coefficient < 0.05 && step_timings.len() >= 3 {
            TimingVerdict::Artificial
        } else if round_number_ratio > 0.5 {
            TimingVerdict::Artificial
        } else if variance_coefficient > 0.1 {
            TimingVerdict::Natural
        } else {
            TimingVerdict::Inconclusive
        };

        TimingPatternAnalysis {
            variance_coefficient: (variance_coefficient * 1000.0).round() / 1000.0,
            trend,
            round_number_ratio: (round_number_ratio * 100.0).round() / 100.0,
            verdict,
        }
    }

    fn make_default_baseline(&self) -> TimingBaseline {
        TimingBaseline {
            challenge_type: "default".into(),
            difficulty: Difficulty::Medium,
            mean_ms: (self.defaults.ai_lower + self.defaults.ai_upper) / 2.0,
            std_ms: (self.defaults.ai_upper - self.defaults.ai_lower) / 4.0,
            too_fast_ms: self.defaults.too_fast,
            ai_lower_ms: self.defaults.ai_lower,
            ai_upper_ms: self.defaults.ai_upper,
            human_ms: self.defaults.human,
            timeout_ms: self.defaults.timeout,
        }
    }

    fn classify_zone(&self, elapsed: f64, baseline: &TimingBaseline) -> TimingZone {
        if elapsed < baseline.too_fast_ms {
            TimingZone::TooFast
        } else if elapsed <= baseline.ai_upper_ms {
            TimingZone::AiZone
        } else if elapsed <= baseline.human_ms {
            TimingZone::Suspicious
        } else if elapsed <= baseline.timeout_ms {
            TimingZone::Human
        } else {
            TimingZone::Timeout
        }
    }

    fn compute_penalty(&self, zone: &TimingZone, elapsed: f64, baseline: &TimingBaseline) -> f64 {
        match zone {
            TimingZone::TooFast => 1.0,
            TimingZone::AiZone => 0.0,
            TimingZone::Suspicious => {
                let range = baseline.human_ms - baseline.ai_upper_ms;
                if range <= 0.0 {
                    return 0.5;
                }
                let position = (elapsed - baseline.ai_upper_ms) / range;
                0.3 + position * 0.4
            }
            TimingZone::Human => 0.9,
            TimingZone::Timeout => 1.0,
        }
    }

    fn compute_z_score(&self, elapsed: f64, baseline: &TimingBaseline) -> f64 {
        if baseline.std_ms == 0.0 {
            return 0.0;
        }
        (elapsed - baseline.mean_ms) / baseline.std_ms
    }

    fn compute_confidence(
        &self,
        elapsed: f64,
        baseline: &TimingBaseline,
        zone: &TimingZone,
    ) -> f64 {
        match zone {
            TimingZone::TooFast => {
                let ratio = elapsed / baseline.too_fast_ms;
                f64::max(0.5, 1.0 - ratio)
            }
            TimingZone::AiZone => {
                let dist = (elapsed - baseline.mean_ms).abs();
                let normalized = dist / baseline.std_ms;
                f64::max(0.5, f64::min(1.0, 1.0 - normalized * 0.15))
            }
            TimingZone::Suspicious => {
                let range = baseline.human_ms - baseline.ai_upper_ms;
                if range <= 0.0 {
                    return 0.5;
                }
                0.4 + 0.2 * ((elapsed - baseline.ai_upper_ms) / range)
            }
            TimingZone::Human => 0.8,
            TimingZone::Timeout => 0.95,
        }
    }

    fn describe_zone(&self, zone: &TimingZone, elapsed: f64, baseline: &TimingBaseline) -> String {
        let ms = elapsed.round() as u64;
        match zone {
            TimingZone::TooFast => format!(
                "Response time {}ms is below {}ms threshold \u{2014} likely pre-computed or scripted",
                ms, baseline.too_fast_ms as u64
            ),
            TimingZone::AiZone => format!(
                "Response time {}ms is within expected AI range [{}ms, {}ms]",
                ms, baseline.ai_lower_ms as u64, baseline.ai_upper_ms as u64
            ),
            TimingZone::Suspicious => format!(
                "Response time {}ms exceeds AI range \u{2014} possible human assistance",
                ms
            ),
            TimingZone::Human => format!(
                "Response time {}ms exceeds {}ms \u{2014} likely human solver",
                ms, baseline.human_ms as u64
            ),
            TimingZone::Timeout => format!(
                "Response time {}ms exceeds timeout threshold of {}ms",
                ms, baseline.timeout_ms as u64
            ),
        }
    }

    fn detect_trend(&self, timings: &[f64]) -> TimingTrend {
        if timings.len() < 3 {
            return TimingTrend::Variable;
        }

        let n = timings.len() as f64;
        let x_mean = (n - 1.0) / 2.0;
        let y_mean: f64 = timings.iter().sum::<f64>() / n;

        let mut numerator = 0.0;
        let mut denominator = 0.0;
        for (i, t) in timings.iter().enumerate() {
            let xi = i as f64;
            numerator += (xi - x_mean) * (t - y_mean);
            denominator += (xi - x_mean).powi(2);
        }

        if denominator == 0.0 {
            return TimingTrend::Constant;
        }

        let slope = numerator / denominator;
        let normalized_slope = if y_mean > 0.0 { slope / y_mean } else { 0.0 };

        if normalized_slope.abs() < 0.05 {
            TimingTrend::Constant
        } else if normalized_slope > 0.1 {
            TimingTrend::Increasing
        } else if normalized_slope < -0.1 {
            TimingTrend::Decreasing
        } else {
            TimingTrend::Variable
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn default_config() -> TimingConfig {
        TimingConfig::default()
    }

    fn analyzer() -> TimingAnalyzer {
        TimingAnalyzer::new(&default_config())
    }

    #[test]
    fn test_zone_too_fast() {
        let a = analyzer();
        let result = a.analyze(5.0, "crypto-nl", &Difficulty::Easy, None);
        assert_eq!(result.zone, TimingZone::TooFast);
        assert_eq!(result.penalty, 1.0);
    }

    #[test]
    fn test_zone_ai() {
        let a = analyzer();
        let result = a.analyze(200.0, "crypto-nl", &Difficulty::Easy, None);
        assert_eq!(result.zone, TimingZone::AiZone);
        assert_eq!(result.penalty, 0.0);
    }

    #[test]
    fn test_zone_suspicious() {
        let a = analyzer();
        let result = a.analyze(4000.0, "crypto-nl", &Difficulty::Easy, None);
        assert_eq!(result.zone, TimingZone::Suspicious);
        assert!(result.penalty > 0.0 && result.penalty < 1.0);
    }

    #[test]
    fn test_zone_human() {
        let a = analyzer();
        let result = a.analyze(12000.0, "crypto-nl", &Difficulty::Easy, None);
        assert_eq!(result.zone, TimingZone::Human);
    }

    #[test]
    fn test_zone_timeout() {
        let a = analyzer();
        let result = a.analyze(35000.0, "crypto-nl", &Difficulty::Easy, None);
        assert_eq!(result.zone, TimingZone::Timeout);
        assert_eq!(result.penalty, 1.0);
    }

    #[test]
    fn test_pattern_natural() {
        let a = analyzer();
        let result = a.analyze_pattern(&[150.0, 230.0, 180.0, 310.0, 190.0]);
        assert_eq!(result.verdict, TimingVerdict::Natural);
    }

    #[test]
    fn test_pattern_artificial() {
        let a = analyzer();
        // Very consistent timings = artificial
        let result = a.analyze_pattern(&[100.0, 100.0, 100.0, 100.0]);
        assert_eq!(result.verdict, TimingVerdict::Artificial);
    }

    #[test]
    fn test_pattern_round_numbers() {
        let a = analyzer();
        // Many round numbers = artificial
        let result = a.analyze_pattern(&[500.0, 1000.0, 500.0, 1000.0]);
        assert_eq!(result.verdict, TimingVerdict::Artificial);
    }
}
